File: multi-processor-programming-study/solutions/cpp/hazard_pointers_solution.cpp
/**
 * SOLUTION: Hazard Pointers for Safe Memory Reclamation
 *
 * Complete reference implementation with all TODOs filled in.
 *
 * This demonstrates safe memory reclamation in lock-free data structures
 * without garbage collection.
 */

#include <iostream>
#include <atomic>
#include <thread>
#include <vector>
#include <array>
#include <memory>

namespace solution {

template <typename T>
class HazardPointerManager {
public:
    static constexpr int MAX_THREADS = 128;
    static constexpr int HAZARD_POINTERS_PER_THREAD = 2;
    static constexpr int RETIRED_LIST_MAX = 100;

private:
    struct HazardPointerRecord {
        std::atomic<T*> hazard_pointer{nullptr};
        std::atomic<bool> active{false};
    };

    std::array<std::array<HazardPointerRecord, HAZARD_POINTERS_PER_THREAD>, MAX_THREADS> hazard_pointers_;
    std::atomic<int> thread_count_{0};

public:
    int acquire_hazard_pointer_slot() {
        int slot = thread_count_.fetch_add(1, std::memory_order_relaxed);
        if (slot >= MAX_THREADS) {
            throw std::runtime_error("Too many threads!");
        }
        return slot;
    }

    void set_hazard_pointer(int thread_id, int hp_index, T* ptr) {
        hazard_pointers_[thread_id][hp_index].hazard_pointer.store(ptr, std::memory_order_release);
        hazard_pointers_[thread_id][hp_index].active.store(true, std::memory_order_release);
    }

    void clear_hazard_pointer(int thread_id, int hp_index) {
        hazard_pointers_[thread_id][hp_index].active.store(false, std::memory_order_release);
        hazard_pointers_[thread_id][hp_index].hazard_pointer.store(nullptr, std::memory_order_release);
    }

    bool is_protected(T* ptr) {
        for (int tid = 0; tid < thread_count_.load(std::memory_order_acquire); ++tid) {
            for (int hp_idx = 0; hp_idx < HAZARD_POINTERS_PER_THREAD; ++hp_idx) {
                if (hazard_pointers_[tid][hp_idx].active.load(std::memory_order_acquire)) {
                    T* hazard = hazard_pointers_[tid][hp_idx].hazard_pointer.load(std::memory_order_acquire);
                    if (hazard == ptr) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    bool try_delete(T* ptr) {
        if (!is_protected(ptr)) {
            delete ptr;
            return true;
        }
        return false;
    }
};

template <typename T>
class HazardPointerContext {
private:
    HazardPointerManager<T>& manager_;
    int thread_id_;
    std::vector<T*> retired_list_;

public:
    HazardPointerContext(HazardPointerManager<T>& manager)
        : manager_(manager), thread_id_(manager.acquire_hazard_pointer_slot()) {}

    T* protect(int hp_index, std::atomic<T*>& atomic_ptr) {
        T* ptr = nullptr;
        T* protected_ptr = atomic_ptr.load(std::memory_order_acquire);

        do {
            ptr = protected_ptr;
            if (ptr == nullptr) {
                manager_.clear_hazard_pointer(thread_id_, hp_index);
                return nullptr;
            }
            manager_.set_hazard_pointer(thread_id_, hp_index, ptr);
            protected_ptr = atomic_ptr.load(std::memory_order_acquire);
        } while (ptr != protected_ptr);

        return ptr;
    }

    void unprotect(int hp_index) {
        manager_.clear_hazard_pointer(thread_id_, hp_index);
    }

    void retire(T* ptr) {
        retired_list_.push_back(ptr);
        if (retired_list_.size() >= HazardPointerManager<T>::RETIRED_LIST_MAX) {
            reclaim();
        }
    }

    void reclaim() {
        std::vector<T*> still_protected;
        for (T* ptr : retired_list_) {
            if (!manager_.try_delete(ptr)) {
                still_protected.push_back(ptr);
            }
        }
        retired_list_ = std::move(still_protected);
    }

    ~HazardPointerContext() {
        reclaim();
        for (T* ptr : retired_list_) {
            delete ptr;
        }
    }
};

template <typename T>
class LockFreeStackWithHP {
private:
    struct Node {
        T value;
        std::atomic<Node*> next;
        explicit Node(const T& val) : value(val), next(nullptr) {}
    };

    std::atomic<Node*> top_{nullptr};
    HazardPointerManager<Node> hp_manager_;
    static thread_local std::unique_ptr<HazardPointerContext<Node>> hp_context_;

    HazardPointerContext<Node>* get_context() {
        if (!hp_context_) {
            hp_context_ = std::make_unique<HazardPointerContext<Node>>(hp_manager_);
        }
        return hp_context_.get();
    }

public:
    void push(const T& value) {
        Node* node = new Node(value);
        Node* old_top = top_.load(std::memory_order_relaxed);

        do {
            node->next.store(old_top, std::memory_order_relaxed);
        } while (!top_.compare_exchange_weak(old_top, node,
                                             std::memory_order_release,
                                             std::memory_order_relaxed));
    }

    bool pop(T& result) {
        auto* context = get_context();
        Node* old_top = nullptr;

        while (true) {
            old_top = context->protect(0, top_);

            if (old_top == nullptr) {
                context->unprotect(0);
                return false;
            }

            Node* next = old_top->next.load(std::memory_order_acquire);

            if (top_.compare_exchange_weak(old_top, next,
                                          std::memory_order_release,
                                          std::memory_order_relaxed)) {
                result = old_top->value;
                context->unprotect(0);
                context->retire(old_top);
                return true;
            }
        }
    }

    ~LockFreeStackWithHP() {
        Node* current = top_.load(std::memory_order_relaxed);
        while (current != nullptr) {
            Node* next = current->next.load(std::memory_order_relaxed);
            delete current;
            current = next;
        }
    }
};

template <typename T>
thread_local std::unique_ptr<HazardPointerContext<typename LockFreeStackWithHP<T>::Node>>
    LockFreeStackWithHP<T>::hp_context_;

void demonstrate_solution() {
    std::cout << "=== SOLUTION: Hazard Pointers ===\n\n";

    std::cout << "Key Points:\n\n";

    std::cout << "1. Protection Protocol:\n";
    std::cout << "   - Set hazard pointer to node\n";
    std::cout << "   - Re-read atomic pointer\n";
    std::cout << "   - If unchanged, node is protected\n";
    std::cout << "   - If changed, retry\n\n";

    std::cout << "2. Retirement:\n";
    std::cout << "   - Add node to retired list\n";
    std::cout << "   - When list is large, scan hazard pointers\n";
    std::cout << "   - Delete unprotected nodes\n\n";

    std::cout << "3. Thread-Local Context:\n";
    std::cout << "   - Each thread has hazard pointer slots\n";
    std::cout << "   - Each thread has retired list\n";
    std::cout << "   - Decouples threads for scalability\n\n";

    // Test
    LockFreeStackWithHP<int> stack;
    stack.push(1);
    stack.push(2);
    stack.push(3);

    int value;
    stack.pop(value);
    std::cout << "Test pop: " << value << " (should be 3)\n";

    std::cout << "\n✅ Solution provides safe memory reclamation\n";
    std::cout << "✅ No use-after-free bugs\n";
    std::cout << "✅ Essential for C++ lock-free programming\n";
}

} // namespace solution

int main() {
    solution::demonstrate_solution();
    return 0;
}

File: multi-processor-programming-study/solutions/java/HazardPointers_Solution.java
/**
 * SOLUTION: Hazard Pointers for Safe Memory Reclamation
 *
 * Complete reference implementation for Java.
 *
 * NOTE: While Java has GC and doesn't strictly need hazard pointers,
 * this implementation is educational for understanding:
 * 1. The memory reclamation problem in C++
 * 2. Cross-language concurrent programming
 * 3. What garbage collection does automatically
 */

package solutions;

import java.util.concurrent.atomic.AtomicReference;
import java.util.concurrent.atomic.AtomicReferenceArray;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.ArrayList;
import java.util.List;

public class HazardPointers_Solution {

    public static class HazardPointerManager<T> {
        private static final int MAX_THREADS = 128;
        private static final int HAZARD_POINTERS_PER_THREAD = 2;
        private static final int RETIRED_LIST_MAX = 100;

        private static class HazardPointerRecord<T> {
            final AtomicReference<T> hazardPointer = new AtomicReference<>(null);
            final AtomicBoolean active = new AtomicBoolean(false);
        }

        private final AtomicReferenceArray<HazardPointerRecord<T>[]> hazardPointers;
        private final AtomicInteger threadCount = new AtomicInteger(0);

        @SuppressWarnings("unchecked")
        public HazardPointerManager() {
            this.hazardPointers = new AtomicReferenceArray<>(MAX_THREADS);
            for (int i = 0; i < MAX_THREADS; i++) {
                HazardPointerRecord<T>[] records = (HazardPointerRecord<T>[])
                    new HazardPointerRecord[HAZARD_POINTERS_PER_THREAD];
                for (int j = 0; j < HAZARD_POINTERS_PER_THREAD; j++) {
                    records[j] = new HazardPointerRecord<>();
                }
                hazardPointers.set(i, records);
            }
        }

        public int acquireHazardPointerSlot() {
            int slot = threadCount.getAndIncrement();
            if (slot >= MAX_THREADS) {
                throw new RuntimeException("Too many threads!");
            }
            return slot;
        }

        public void setHazardPointer(int threadId, int hpIndex, T ptr) {
            HazardPointerRecord<T>[] records = hazardPointers.get(threadId);
            records[hpIndex].hazardPointer.set(ptr);
            records[hpIndex].active.set(true);
        }

        public void clearHazardPointer(int threadId, int hpIndex) {
            HazardPointerRecord<T>[] records = hazardPointers.get(threadId);
            records[hpIndex].active.set(false);
            records[hpIndex].hazardPointer.set(null);
        }

        public boolean isProtected(T ptr) {
            int count = threadCount.get();
            for (int tid = 0; tid < count; tid++) {
                HazardPointerRecord<T>[] records = hazardPointers.get(tid);
                for (int hpIdx = 0; hpIdx < HAZARD_POINTERS_PER_THREAD; hpIdx++) {
                    if (records[hpIdx].active.get()) {
                        T hazard = records[hpIdx].hazardPointer.get();
                        if (hazard == ptr) {
                            return true;
                        }
                    }
                }
            }
            return false;
        }

        public boolean tryReclaim(T ptr) {
            return !isProtected(ptr);
        }
    }

    public static class HazardPointerContext<T> {
        private final HazardPointerManager<T> manager;
        private final int threadId;
        private final List<T> retiredList = new ArrayList<>();

        public HazardPointerContext(HazardPointerManager<T> manager) {
            this.manager = manager;
            this.threadId = manager.acquireHazardPointerSlot();
        }

        public T protect(int hpIndex, AtomicReference<T> atomicRef) {
            T ptr = null;
            T protectedPtr = atomicRef.get();

            do {
                ptr = protectedPtr;
                if (ptr == null) {
                    manager.clearHazardPointer(threadId, hpIndex);
                    return null;
                }
                manager.setHazardPointer(threadId, hpIndex, ptr);
                protectedPtr = atomicRef.get();
            } while (ptr != protectedPtr);

            return ptr;
        }

        public void unprotect(int hpIndex) {
            manager.clearHazardPointer(threadId, hpIndex);
        }

        public void retire(T ptr) {
            retiredList.add(ptr);
            if (retiredList.size() >= HazardPointerManager.RETIRED_LIST_MAX) {
                reclaim();
            }
        }

        public void reclaim() {
            List<T> stillProtected = new ArrayList<>();
            for (T ptr : retiredList) {
                if (!manager.tryReclaim(ptr)) {
                    stillProtected.add(ptr);
                }
            }
            retiredList.clear();
            retiredList.addAll(stillProtected);
        }
    }

    public static class LockFreeStackWithHP<T> {
        private static class Node<T> {
            final T value;
            final AtomicReference<Node<T>> next;

            Node(T value) {
                this.value = value;
                this.next = new AtomicReference<>(null);
            }
        }

        private final AtomicReference<Node<T>> top = new AtomicReference<>(null);
        private final HazardPointerManager<Node<T>> hpManager = new HazardPointerManager<>();
        private final ThreadLocal<HazardPointerContext<Node<T>>> hpContext =
            ThreadLocal.withInitial(() -> new HazardPointerContext<>(hpManager));

        public void push(T value) {
            Node<T> node = new Node<>(value);
            Node<T> oldTop;

            do {
                oldTop = top.get();
                node.next.set(oldTop);
            } while (!top.compareAndSet(oldTop, node));
        }

        public T pop() {
            HazardPointerContext<Node<T>> context = hpContext.get();
            Node<T> oldTop;

            while (true) {
                oldTop = context.protect(0, top);

                if (oldTop == null) {
                    context.unprotect(0);
                    return null;
                }

                Node<T> next = oldTop.next.get();

                if (top.compareAndSet(oldTop, next)) {
                    T value = oldTop.value;
                    context.unprotect(0);
                    context.retire(oldTop);
                    return value;
                }
            }
        }
    }

    public static void main(String[] args) {
        System.out.println("=== SOLUTION: Hazard Pointers ===\n");

        System.out.println("Key Implementation Points:\n");

        System.out.println("1. Protection Protocol:");
        System.out.println("   - Set hazard pointer to target node");
        System.out.println("   - Re-read atomic reference");
        System.out.println("   - If unchanged, protection successful");
        System.out.println("   - If changed, retry loop\n");

        System.out.println("2. Retirement and Reclamation:");
        System.out.println("   - Removed nodes go to retired list");
        System.out.println("   - When list grows, scan hazard pointers");
        System.out.println("   - Reclaim nodes not protected by any thread");
        System.out.println("   - Keep protected nodes in retired list\n");

        System.out.println("3. Thread-Local Design:");
        System.out.println("   - Each thread has its own hazard pointer slots");
        System.out.println("   - Each thread has its own retired list");
        System.out.println("   - Minimizes contention between threads\n");

        // Test
        LockFreeStackWithHP<Integer> stack = new LockFreeStackWithHP<>();
        stack.push(1);
        stack.push(2);
        stack.push(3);

        Integer value = stack.pop();
        System.out.println("Test pop: " + value + " (should be 3)\n");

        System.out.println("✅ Solution implements safe hazard pointer protocol");
        System.out.println("✅ In Java: Educational (GC handles this)");
        System.out.println("✅ In C++: CRITICAL for production lock-free code");
        System.out.println("✅ Cross-language understanding essential");
    }
}
